lib/juttle-service-client.js

```diff
@@ -0,0 +1,347 @@
+                                                  modules: err_obj.info.bundle.modules,
+                                                  filename: filename}));
+    } else {
+        write_error(sprintf('ERROR %s', err.message));
+    }
+}
+
+function read_inputs(opts) {
+
+    let inputs = {};
+    if (opts.input !== undefined) {
+        if (! _.isArray(opts.input)) {
+            opts.input = [opts.input];
+        }
+        _.each(opts.input, function(x) {
+            let L = x.split('=');
+            if (L.length !== 2) {
+                write_error(sprintf('error: invalid input %s', x));
+                usage();
+            }
+            let num = +L[1];
+            inputs[L[0]] = isNaN(num) ? L[1] : num;
+        });
+    }
+
+    return inputs;
+}
+
+let cmdargs = {
+    'path': {
+        usage: '--path <path-to-juttle-file>',
+        desc: 'Path to file relative to configured root directory.'
+    },
+    'wait': {
+        usage: '--wait',
+        desc: 'If true, wait for program to finish, otherwise run in background'
+    },
+    'input': {
+        usage: '--input name=val',
+        desc: 'One or more input values.'
+    },
+    'job': {
+        usage: '--job <job-id>',
+        desc: 'Job id.'
+    },
+    'observer': {
+        usage: '--observer <observer-id>',
+        desc: 'Observer id.'
+    },
+    'server': {
+        usage: '--juttle-service <hostname:port>',
+        desc: 'Hostname/port of server'
+    },
+    help: {
+        usage: '--help',
+        desc: 'Print this help and exit'
+    }
+}
+
+let commands = {
+    subscribe: {
+        usage: '({job} | {observer})',
+        perform: subscribe
+    },
+    list_jobs: {
+        usage: '[{job}]',
+        perform: list_jobs
+    },
+    list_observers: {
+        usage: '',
+        perform: list_observers
+    },
+    run: {
+        usage: '{path} [{wait}] [{observer}]',
+        perform: run
+    },
+    delete: {
+        usage: '{job}',
+        perform: delete_job
+    },
+    get_inputs: {
+        usage: '{path} {input} [{input} ...]',
+        perform: get_inputs
+    }
+};
+
+function expand(string) {
+    return string.replace(/\{([^}]*)\}/g, function(match, key) {
+        return cmdargs[key].usage;
+    });
+}
+
+function init(options) {
+    write_output = options.write_output;
+    write_error = options.write_error;
+    client_exit = options.client_exit;
+    commands = _.extend(commands, options.commands);
+    cmdargs = _.extend(cmdargs, options.cmdargs);
+}
+
+function usage() {
+
+    // Given the command descriptions, create the list of commands
+    // used by each option.
+    _.each(cmdargs, function (arginfo, arg) {
+        arginfo.commands = [];
+        _.each(commands, function(info, cmd) {
+            if (info.usage.indexOf('{' + arg + '}') > -1) {
+                arginfo.commands.push(cmd);
+            }
+        });
+    });
+
+    // Expand each command usage
```

Comment:
This is nice and DRY but is it really worth it? I'd find the code more readable if each command's usage just listed its options (repeating if necessary) rather than this expansion stuff.